### PR TITLE
feature: Add pytorch lighting optimized text classification pipeline

### PR DESCRIPTION
**Pull request by ktagowski**

### CODE REVIEW

**Comment by ktagowski** [comment]
TODO:
- [ ]  Fix weird mypy error
- [ ]  Correct ConfigSpace parameters like max_epochs etc.
- [ ] Check pipeline on GPU

File: embeddings/hyperparameter_search/lighting_configspace.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    weight_decay: Parameter = SearchableParameter(
+        name="weight_decay",
+        type="uniform",
+        low=0.0,
+        high=1e-1,
+    )
+    finetune_last_n_layers: Parameter = SearchableParameter(
+        name="finetune_last_n_layers",
+        type="categorical",
+        choices=[-1, 4, 7, 9, 11, None],
```
Comment (djaniak) [inline embeddings/hyperparameter_search/lighting_configspace.py:58]: finetune_last_n_layers cannot be None

Comment (djaniak) [reply embeddings/hyperparameter_search/lighting_configspace.py:58]:
```suggestion
        choices=[1, 3, 5, 9, -1],
```

File: embeddings/model/lightning_model.py [hunk 2]
```diff
@@ -21,7 +22,13 @@ def __init__(
         self.predict_subset = predict_subset
 
     def execute(self, data: pl.LightningDataModule) -> Dict[str, nptyping.NDArray[Any]]:
-        self.trainer.fit(self.task, data)
+        try:
+            self.trainer.fit(self.task, data)
+        except Exception as e:
+            del self.trainer
```
Comment (djaniak) [inline embeddings/model/lightning_model.py:28]: should this be put here or somewhere in hps related code? would teardown method work here as well?

**Comment by djaniak** [comment]
what about moving flair related config space to seperate file (as lightning config spaces are in related one)? finally we woudl have 3 files configspace, flairconfigspace and lightningcofnigspace

File: embeddings/model/lightning_model.py (hunk 2 above)
Comment (ktagowski) [reply embeddings/model/lightning_model.py:28]: It should be placed here as error while training may occur also while not hps related code. As I understand teardown correctly it's called every time after fit, not only while error occurs.

**Comment by ktagowski** [comment]
> what about moving flair related config space to seperate file (as lightning config spaces are in related one)? finally we woudl have 3 files configspace, flairconfigspace and lightningcofnigspace

Done

File: embeddings/data/datamodule.py [hunk 3]
```diff
@@ -37,7 +41,7 @@ class HuggingFaceDataModule(BaseDataModule[DatasetDict]):
     def __init__(
         self,
         tokenizer_name_or_path: str,
-        dataset_name: str,
+        dataset_name_or_path: Union[str, "os.PathLike[Any]"],
```
Comment (laugustyniak) [inline embeddings/data/datamodule.py:44]: why not?
```suggestion
        dataset_name_or_path: Union[str, pathlib.Path],
```

File: embeddings/data/datamodule.py [hunk 4]
```diff
@@ -71,10 +75,18 @@ def __init__(
         )
         self.load_dataset_kwargs = load_dataset_kwargs if load_dataset_kwargs else {}
 
-    def load_dataset(self) -> DatasetDict:
-        result = datasets.load_dataset(self.dataset_name, **self.load_dataset_kwargs)
-        assert isinstance(result, DatasetDict)
-        return result
+    def load_dataset(self, preparation_step: bool = False) -> DatasetDict:
+        loader: Union[HuggingFaceDataLoader, HuggingFaceLocalDataLoader] = HuggingFaceDataLoader()
+        if os.path.exists(self.dataset_name_or_path):
```
Comment (laugustyniak) [inline embeddings/data/datamodule.py:80]:
```suggestion
        if exists(self.dataset_name_or_path):
```

and correct import accordingly

File: embeddings/pipeline/lightning_classification.py [hunk 5]
```diff
@@ -35,10 +37,11 @@ def __init__(
         load_dataset_kwargs: Optional[Dict[str, Any]] = None,
         task_model_kwargs: Optional[Dict[str, Any]] = None,
         task_train_kwargs: Optional[Dict[str, Any]] = None,
+        predict_subset: Literal["dev", "test"] = "test",
```
Comment (laugustyniak) [inline embeddings/pipeline/lightning_classification.py:40]: do we want to create an enum for sub-set names? `dev` / `val` / `validation` will not be mixed in future  :D

File: embeddings/transformation/hf_transformation/sampling_transformation.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+
+class SampleSplitsHuggingFaceTransformation(
+    Transformation[datasets.DatasetDict, datasets.DatasetDict]
+):
+    def __init__(
+        self,
+        dev_fraction: Optional[float] = None,
+        test_fraction: Optional[float] = None,
+        seed: int = 441,
```
Comment (laugustyniak) [inline embeddings/transformation/hf_transformation/sampling_transformation.py:15]: great seed :D

File: embeddings/transformation/hf_transformation/sampling_transformation.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            sampled_dataset = self._train_test_split(data["train"], self.dev_fraction)
+            dataset["train"] = sampled_dataset["train"]
+            dataset["validation"] = sampled_dataset["test"]
+
+        elif self.test_fraction and "test" not in data:
+            dataset = self._train_test_split(data["train"], self.test_fraction)
+
+        if not dataset:
+            dataset = data
+        return dataset
```
Comment (laugustyniak) [inline embeddings/transformation/hf_transformation/sampling_transformation.py:59]:
```suggestion

        else:
            return data
        return dataset
```

Comment (laugustyniak) [reply embeddings/transformation/hf_transformation/sampling_transformation.py:59]: maybe this?

Comment (laugustyniak) [reply embeddings/transformation/hf_transformation/sampling_transformation.py:59]: or other if checks could return an empty dataset?

File: embeddings/pipeline/lightning_classification.py (hunk 5 above)
Comment (ktagowski) [reply embeddings/pipeline/lightning_classification.py]: Done

**Comment by ktagowski** [comment]
Mypy issue moved to separate GH issue: https://github.com/CLARIN-PL/embeddings/issues/152

File: embeddings/hyperparameter_search/lighting_configspace.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    trainer_devices: Parameter = field(init=False)
+    trainer_accelerator: Parameter = field(init=False)
+
+    max_epochs: Parameter = SearchableParameter(
+        name="max_epochs", type="int_uniform", low=1, high=30, step=1
+    )
+    mini_batch_size: Parameter = SearchableParameter(
+        name="batch_size", type="log_int_uniform", low=8, high=64, step=1
+    )
+    max_seq_length: Parameter = ConstantParameter(
```
Comment (djaniak) [inline embeddings/hyperparameter_search/lighting_configspace.py:32]: why is this constant parameter?

File: embeddings/hyperparameter_search/lighting_configspace.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        task_model_kwargs = BaseConfigSpace._pop_parameters(
+            parameters=parameters, parameters_keys=task_model_keys
+        )
+        task_train_kwargs = BaseConfigSpace._pop_parameters(
+            parameters=parameters, parameters_keys=task_train_keys
+        )
+
+        batch_size = pipeline_kwargs.pop("batch_size")
+        pipeline_kwargs["train_batch_size"] = batch_size
+        pipeline_kwargs["eval_batch_size"] = batch_size
```
Comment (djaniak) [inline embeddings/hyperparameter_search/lighting_configspace.py:117]: maybe simplify with chained assignment?
```suggestion
        pipeline_kwargs["train_batch_size"] = pipeline_kwargs["eval_batch_size"]  = batch_size
```

File: embeddings/model/lightning_model.py [hunk 10]
```diff
@@ -13,17 +15,21 @@ def __init__(
         self,
         trainer: pl.Trainer,
         task: HuggingFaceLightningTask,
-        predict_subset: Literal["dev", "test"] = "test",
+        predict_subset: LightingDataModuleSubset,
```
Comment (djaniak) [inline embeddings/model/lightning_model.py:18]: predict dataloader is not defined in datamodule although we allow to use it for predicting; it also does not make sense to predict on train dataloader

File: embeddings/pipeline/lighting_hps_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+class _OptimizedLightingPipelineBase(
+    _HuggingFaceOptimizedPipelineBase[ConfigSpace], ABC, Generic[ConfigSpace]
+):
+    input_column_name: str
+    target_column_name: str
+
+    dataset_dir: TemporaryDirectory[str] = field(init=False, default_factory=TemporaryDirectory)
+    tmp_model_output_dir: TemporaryDirectory[str] = field(
+        init=False, default_factory=TemporaryDirectory
+    )
```
Comment (djaniak) [inline embeddings/pipeline/lighting_hps_pipeline.py:33]: both fields are temporary; should we unify the names? e.g. remove tmp_ from tmp_model_output_dir to add to dataset_dir

File: embeddings/model/lightning_model.py (hunk 10 above)
Comment (ktagowski) [reply embeddings/model/lightning_model.py:18]: It's universal class, predict data loader should be allowed although

File: embeddings/hyperparameter_search/lighting_configspace.py (hunk 8 above)
Comment (ktagowski) [reply embeddings/hyperparameter_search/lighting_configspace.py:32]: To limit space for search, It can be overridden by the user if otherwise.

File: embeddings/model/lightning_model.py (hunk 10 above)
Comment (ktagowski) [reply embeddings/model/lightning_model.py:18]: Currently we don't support cases whether we want to  skip prediction step, so passing train data loader can be used for such case.

File: embeddings/pipeline/lighting_hps_pipeline.py (hunk 11 above)
Comment (ktagowski) [reply embeddings/pipeline/lighting_hps_pipeline.py:33]: I can add `tmp` to dataset_dir as this fields are `temporal`.  I will also change it in the flair hps pipeline

File: embeddings/model/lightning_model.py (hunk 10 above)
Comment (djaniak) [reply embeddings/model/lightning_model.py:18]: ok

File: embeddings/model/lightning_model.py [hunk 12]
```diff
@@ -13,17 +15,21 @@ def __init__(
...
-            data.test_dataloader() if self.predict_subset == "test" else data.val_dataloader()
-        )
+        try:
+            self.trainer.fit(self.task, data)
+        except Exception as e:
+            del self.trainer
+            torch.cuda.empty_cache()  # type: ignore
+            raise e
+
+        dataloader = get_subset_from_lighting_datamodule(data=data, subset=self.predict_subset)
```
Comment (djaniak) [inline embeddings/model/lightning_model.py:33]: maybe move this functionality to datamodule? 
```suggestion
        dataloader = data.get_subset(self.predict_subset)
```

File: embeddings/task/lightning_task/text_classification.py [hunk 13]
```diff
@@ -105,9 +105,11 @@ def test_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
     def predict(
         self, dataloader: DataLoader[HuggingFaceDataset]
     ) -> Dict[str, nptyping.NDArray[Any]]:
-        predictions = torch.argmax(
-            torch.cat([self.forward(**batch).logits for batch in dataloader]), dim=1
-        ).numpy()
+        assert self.trainer is not None
```
Comment (djaniak) [inline embeddings/task/lightning_task/text_classification.py:108]: how about encapsulating this postprocessing in predict_step method?
https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.core.lightning.html#pytorch_lightning.core.lightning.LightningModule.predict_step

File: embeddings/pipeline/lighting_hps_pipeline.py (hunk 11 above)
Comment (ktagowski) [reply embeddings/pipeline/lighting_hps_pipeline.py]: Done

File: embeddings/model/lightning_model.py (hunk 12 above)
Comment (ktagowski) [reply embeddings/model/lightning_model.py]: Done

File: embeddings/task/lightning_task/text_classification.py (hunk 13 above)
Comment (ktagowski) [reply embeddings/task/lightning_task/text_classification.py:117]: Done

**Review by asawczyn — CHANGES_REQUESTED** [review]

File: embeddings/hyperparameter_search/lighting_configspace.py (hunk 8 above)
Comment (djaniak) [reply embeddings/hyperparameter_search/lighting_configspace.py:32]: Ok I see, but I'd still make this one a SearchableParameter, just with fewer values, and limit the pool of other parameters like finetune_last_n_layers. Setting only this one parameter as constant does not seem to be consistent.

File: embeddings/data/datamodule.py [hunk 14]
```diff
@@ -70,11 +76,20 @@ def __init__(
...
-
-    def load_dataset(self) -> DatasetDict:
-        result = datasets.load_dataset(self.dataset_name, **self.load_dataset_kwargs)
-        assert isinstance(result, DatasetDict)
-        return result
+        self.seed = seed
+
+    def load_dataset(self, preparation_step: bool = False) -> DatasetDict:
+        loader: Union[HuggingFaceDataLoader, HuggingFaceLocalDataLoader] = HuggingFaceDataLoader()
+        if exists(self.dataset_name_or_path):
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/data/datamodule.py:84]: Added `NotImplementedError` exception when path is not dir, so it should cover all cases now

File: embeddings/pipeline/lighting_hps_pipeline.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+        Dict[str, ParameterValues],
+    ]:
+        embedding_name = parameters["embedding_name"]
+        assert isinstance(embedding_name, str)
+        train_batch_size = parameters["train_batch_size"]
+        assert isinstance(train_batch_size, int)
+        eval_batch_size = parameters["eval_batch_size"]
+        assert isinstance(eval_batch_size, int)
+        finetune_last_n_layers = parameters["finetune_last_n_layers"]
+        assert finetune_last_n_layers is None or isinstance(finetune_last_n_layers, int)
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/pipeline/lighting_hps_pipeline.py]: Fixed, in previous versions I have used None.

File: tests/test_hf_preprocessing_pipeline.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+        "dataset_name": hf_dataset_cfg["dataset"],
+        "persist_path": result_path.name,
+        "sample_missing_splits": None,
+        "ignore_test_subset": True,
+        "seed": 441,
+        "load_dataset_kwargs": None,
+    }
+
+
+@pytest.fixture
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply tests/test_hf_preprocessing_pipeline.py]: Added

File: embeddings/transformation/hf_transformation/sampling_transformation.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+        elif self.dev_fraction and "validation" not in data:
+            sampled_dataset = self._train_test_split(data["train"], self.dev_fraction)
+            dataset["train"] = sampled_dataset["train"]
+            dataset["validation"] = sampled_dataset["test"]
+
+        elif self.test_fraction and "test" not in data:
+            dataset = self._train_test_split(data["train"], self.test_fraction)
+
+        else:
+            return data
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/transformation/hf_transformation/sampling_transformation.py:58]: We can pass additional flag `replace_original_splits` if set to True splits will be replaced, in other case it Raises Warning. WDYT?

File: embeddings/hyperparameter_search/lighting_configspace.py (hunk 8 above)
Comment (ktagowski) [reply embeddings/hyperparameter_search/lighting_configspace.py:32]: Why? We have limited resources to fine tune all parameters, so only part of it should be tuned. I would update configspaces after performing experiments to choose best performing ones based on the results.

File: embeddings/transformation/hf_transformation/sampling_transformation.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (djaniak) [reply embeddings/transformation/hf_transformation/sampling_transformation.py:58]: I think we only want to sample if the given split is missing, otherwise do not perform the transformation, which is correct with the logic. I guess we could inform the user whether or not the given split is not missing.
In case of `self.dev_fraction and self.test_fraction and "validation" not in data.keys() and "test" in data.keys())` it will sample the validation split

File: tests/test_hf_preprocessing_pipeline.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ktagowski) [reply tests/test_hf_preprocessing_pipeline.py]: Fixed

File: embeddings/transformation/hf_transformation/sampling_transformation.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/transformation/hf_transformation/sampling_transformation.py:74]: I added warnings. @asawczyn @djaniak please check whether it solve your concerns

File: embeddings/hyperparameter_search/lighting_configspace.py (hunk 8 above)
Comment (ktagowski) [reply embeddings/hyperparameter_search/lighting_configspace.py:32]: I would transfer all concerns with ConfigSpace setup to separate issue, after the experiments will be finished. WDYT? @djaniak @asawczyn From my point of view it can be any value set here as for now.

File: embeddings/transformation/hf_transformation/sampling_transformation.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (asawczyn) [reply embeddings/transformation/hf_transformation/sampling_transformation.py:74]: I would like a more informative warning. `"Sampling transformation wrongly defined. Subsets can not be overwritten. Returning original dataset."`

[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/transformation/hf_transformation/sampling_transformation.py:77]: Changed

**Review by djaniak — APPROVED** [review]

File: embeddings/hyperparameter_search/lighting_configspace.py (hunk 8 above)
Comment (djaniak) [reply embeddings/hyperparameter_search/lighting_configspace.py:32]: Ok let's do that, here i don't like that we cherry picking this specific parameter to not tune, especially that we pick value for which the training takes the longest

**Review by asawczyn — APPROVED** [review]
